hrpc/snapshot.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (C) 2015  The GoHBase Authors.  All rights reserved.
+// This file is part of GoHBase.
+// Use of this source code is governed by the Apache License 2.0
+// that can be found in the COPYING file.
+package hrpc
+
+import (
+	"context"
+	"errors"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/tsuna/gohbase/pb"
+)
+
+type snap struct {
+	name         string
+	table        string
+	snapshotType *pb.SnapshotDescription_Type
+	version      int32
+	owner        string
+}
+
+func (s *snap) ToProto() *pb.SnapshotDescription {
+	return &pb.SnapshotDescription{
+		Type:    s.snapshotType,
+		Table:   proto.String(s.table),
+		Name:    proto.String(s.name),
+		Version: proto.Int32(s.version),
+		Owner:   proto.String(s.owner),
+	}
+}
+
+func (s *snap) Version(v int32) {
+	s.version = v
+}
+
+type SnapshotSettable interface {
+	Version(int32)
+}
+
+func SnaphotVersion(v int32) func(Call) error {
+	return func(g Call) error {
+		sn, ok := g.(SnapshotSettable)
+		if !ok {
+			return errors.New("'SnapshotVersion' option can only be used with Snapshot queries")
+		}
+		sn.Version(v)
+		return nil
+	}
+}
+
+// Snapshot represents a Snapshot HBase call
+type Snapshot struct {
+	base
+	snap
+}
+
+// NewSnapshot creates a new Snapshot request that will request a
+// new snapshot in HBase.
+func NewSnapshot(ctx context.Context, name string, table string) *Snapshot {
+	return &Snapshot{
+		base{
+			table:    []byte(table),
+			ctx:      ctx,
+			resultch: make(chan RPCResult, 1),
+		},
+		snap{
+			name: name, table: table,
+		},
+	}
+}
+
+// SetContext sets a new context for the Snapshot.
+//
+// This is used to reset the context deadline in cases where we need to cleanup
+// by deleting a snapshot if creating that snapshot timed out.
+func (sr *Snapshot) SetContext(ctx context.Context) {
+	sr.base.ctx = ctx
+}
+
+// Name returns the name of this RPC call.
+func (sr *Snapshot) Name() string {
+	return "Snapshot"
+}
+
+// ToProto converts the RPC into a protobuf message
+func (sr *Snapshot) ToProto() proto.Message {
+	return &pb.SnapshotRequest{Snapshot: sr.snap.ToProto()}
+}
+
+// NewResponse creates an empty protobuf message to read the response of this
+// RPC.
+func (sr *Snapshot) NewResponse() proto.Message {
+	return &pb.SnapshotResponse{}
+}
+
+// CheckSnapshot represents a CheckSnapshot HBase call
+type CheckSnapshot struct {
+	*Snapshot
+}
+
+// NewCheckSnapshot creates a new CheckSnapshot request that will request a
+// new snapshot in HBase.
+func NewCheckSnapshot(t *Snapshot) *CheckSnapshot {
```

Comment:
could we call it consistently with service name "IsSnapshotDone"? CheckSnapshot seems like a more generic name which assume that it validates that snapshot is intact.